### PR TITLE
RDM -1001 Upgrade Case Doc AM Code Latest Versions/Standards/Norms 

### PR DESCRIPTION
**Pull request by smathangi**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-11309

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: .github/workflows/swagger.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+          restore-keys: ${{ runner.os }}-gradle-
+      - name: Set up JDK 11
+        uses: actions/setup-java@v1
+        with:
+          java-version: 11
+      - name: Run Swagger Publisher
+        run: ./gradlew integration --tests uk.gov.hmcts.reform.ccd.documentam.config.SwaggerPublisher
+      - name: Commit to repository
+        run: |
+          mkdir swagger-staging
```
Comment (sullivangareth) [inline .github/workflows/swagger.yml:27]: Is there a scenario where this directory may already exist ?

 `mkdir -p swagger-staging` so it doesn't error if dir already exists?

Comment (sullivangareth) [reply .github/workflows/swagger.yml:27]: or we could delete the directory at the end ?

File: build.gradle [hunk 2]
```diff
@@ -213,39 +193,19 @@ sonarqube {
...
...perty "sonar.exclusions", "src/main/java/uk/gov/hmcts/reform/ccd/document/am/*Application.java," + "src/main/java/uk/gov/hmcts/reform/ccd/document/am/util/*Swagger2SpringBoot.java," + "src/main/java/uk/gov/hmcts/reform/ccd/document/am/controller/advice/*ErrorConstants.java," + "src/main/java/uk/gov/hmcts/reform/ccd/document/am/controller/advice/exception/**," + "src/main/java/uk/gov/hmcts/reform/ccd/document/am/configuration/**," + "src/main/java/uk/gov/hmcts/reform/ccd/document/am/model/enums/**," + "src/main/java/uk/gov/hmcts/reform/ccd/document/am/apihelper/**," + "src/main/java/uk/gov/hmcts/reform/ccd/documentam/service/impl/DocumentManagementServiceImpl.java," + "src/main/java/uk/gov/hmcts/reform/ccd/documentam/model/StoredDocumentHalResource.java," + "src/main/java/uk/gov/hmcts/reform/ccd/documentam/util/ResponseHelper.java," + "src/main/java/uk/gov/hmcts/reform/ccd/documentam/apihelper/Constants.java," + "src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/endpoints/CaseDocumentAmController.java" + "src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/*.java,src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/endpoints/CaseDocumentAmController.java"
```
Comment (sullivangareth) [inline build.gradle:197]: some of the formatting is a little out on last two lines - keep consistent with one exclusion per line ?

File: build.gradle [hunk 3]
```diff
@@ -273,85 +233,73 @@ repositories {
...
-    exclude group: 'uk.gov.hmcts.reform', module: 'java-logging-spring'
+  implementation group: 'uk.gov.hmcts.reform', name: 'logging', version: versions.reformLogging
+  implementation group: 'uk.gov.hmcts.reform', name: 'logging-appinsights', version: versions.reformLogging
 
-  }
-  compile group: 'javax.inject', name: 'javax.inject', version: '1'
-  compile group: 'io.springfox', name: 'springfox-swagger2', version: versions.swagger2Version
-  compile group: 'io.springfox', name: 'springfox-swagger-ui', version: versions.swagger2Version
-  compile group: 'io.springfox', name: 'springfox-spring-webmvc', version: versions.swagger2Version
+  implementation group: 'uk.gov.hmcts.reform', name: 'service-auth-provider-client', version: '3.1.2'
```
Comment (sullivangareth) [inline build.gradle:257]: should the version be specified in `versions` and referenced the same as other dependencies do ? 

e.g

`versions.serviceAuthProviderClient`

File: build.gradle [hunk 4]
```diff
@@ -273,85 +233,73 @@ repositories {
...
-  compile group: 'uk.gov.hmcts.reform', name: 'logging', version: versions.reformLogging
-  compile group: 'uk.gov.hmcts.reform', name: 'logging-appinsights', version: versions.reformLogging
-  compile group: 'org.springframework.cloud', name: 'spring-cloud-starter-netflix-hystrix', version: '2.2.2.RELEASE'
-  compile('org.springframework.boot:spring-boot-starter-hateoas')
-  compile group: 'org.springframework', name: 'spring-tx', version: '5.2.9.RELEASE'
+  implementation "org.springframework.security:spring-security-web"
+  implementation "org.springframework.security:spring-security-config"
+  implementation "org.springframework.boot:spring-boot-starter-oauth2-client"
+  implementation "org.springframework.boot:spring-boot-starter-oauth2-resource-server"
+  implementation "com.nimbusds:nimbus-jose-jwt:9.6"
```
Comment (sullivangareth) [inline build.gradle:264]: Another convention for specifying a dependency being used here - better to have consistent approach ?

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/apihelper/Constants.java [hunk 5]
```diff
@@ -2,9 +2,6 @@
 
 public class Constants {
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/ccd/documentam/apihelper/Constants.java:3]: Could this class be marked `final` ?

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/advice/CaseDocumentControllerAdvice.java [hunk 6]
```diff
@@ -9,91 +9,86 @@
...
 
-
     @ExceptionHandler(UnauthorizedException.class)
-    protected ResponseEntity<Object> handleUnautorizedExceptionException(
-        HttpServletRequest request,
-        UnauthorizedException exeception
-    ) {
-        return errorDetailsResponseEntity(exeception, HttpStatus.UNAUTHORIZED,
-                                          ErrorConstants.UNAUTHORIZED.getErrorCode(), ErrorConstants.UNAUTHORIZED.getErrorMessage());
+    protected ResponseEntity<Object> handleUnAutorizedExceptionException(UnauthorizedException exception) {
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/advice/CaseDocumentControllerAdvice.java:33]: small typo - wouldn't `handleUnauthorizedException` be a better name ?

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/advice/CaseDocumentControllerAdvice.java [hunk 7]
```diff
@@ -9,91 +9,86 @@
...
+        return errorDetailsResponseEntity(exception,
+                                          HttpStatus.UNAUTHORIZED,
+                                          ErrorConstants.UNAUTHORIZED.getErrorCode(),
+                                          ErrorConstants.UNAUTHORIZED.getErrorMessage()
+        );
+    }
+
+    @ExceptionHandler(ForbiddenException.class)
+    protected ResponseEntity<Object> handleUnAutorizedExceptionException(ForbiddenException exception) {
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/advice/CaseDocumentControllerAdvice.java:42]: change method name to `handleForbiddenException`

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/advice/CaseDocumentControllerAdvice.java [hunk 8]
```diff
@@ -9,91 +9,86 @@
...
+        return errorDetailsResponseEntity(exception, HttpStatus.FORBIDDEN,
+                                          ErrorConstants.ACCESS_DENIED.getErrorCode(),
+                                          ErrorConstants.ACCESS_DENIED.getErrorMessage()
+        );
+    }
+
+    @ExceptionHandler(BadRequestException.class)
+    protected ResponseEntity<Object> badRequestException(BadRequestException ex) {
```
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/ccd/documentam/controller/advice/CaseDocumentControllerAdvice.java:50]: rename to `handleBadRequestException` for consistency ?

File: src/main/resources/application.yaml [hunk 9]
```diff
@@ -22,42 +27,48 @@ spring:
...
-
-  authorisedRoles:
-
-### dependency configuration
-auth.provider.service.client.baseUrl: ${IDAM_S2S_URL:http://localhost:4502}
 idam:
+  api:
+    url: ${IDAM_API_URL:http://localhost:5000}
+    # same for all environments - not used in mca but configured as idam mandates for token generation
+    redirect_uri: https://manage-case-assignment/oauth2redirect
```
Comment (sullivangareth) [inline src/main/resources/application.yaml:58]: mca ?

File: src/test/java/uk/gov/hmcts/reform/ccd/documentam/controller/endpoints/S2SAuthenticationTest.java [hunk 10]
```diff
@@ -49,6 +50,7 @@ public void setUp() {
     }
 
     @Test
+    @Ignore // doing nothing - probably delete??
```
Comment (sullivangareth) [inline src/test/java/uk/gov/hmcts/reform/ccd/documentam/controller/endpoints/S2SAuthenticationTest.java:53]: I'd rather delete - its still in github if we need it ?

File: src/test/java/uk/gov/hmcts/reform/ccd/documentam/security/SecurityUtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+import java.util.Collection;
+import java.util.Date;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import uk.gov.hmcts.reform.authorisation.generators.AuthTokenGenerator;
+import uk.gov.hmcts.reform.idam.client.models.UserInfo;
+
+@DisplayName("SecurityUtils")
+class SecurityUtilsTest {
```
Comment (sullivangareth) [inline src/test/java/uk/gov/hmcts/reform/ccd/documentam/security/SecurityUtilsTest.java:40]: Does this not also require `@ExtendWith(MockitoExtension.class)`?

File: .github/workflows/swagger.yml (hunk 1 above)
Comment (smathangi) [reply .github/workflows/swagger.yml:27]: github actions build creates a new workspace for every build and does clean up as well.

File: build.gradle (hunk 2 above)
Comment (smathangi) [reply build.gradle:197]: I've tried multi line... but it didn't work. Will see if i can format in a better way.

File: .github/workflows/swagger.yml (hunk 1 above)
Comment (sullivangareth) [reply .github/workflows/swagger.yml:27]: ok great !

File: src/test/java/uk/gov/hmcts/reform/ccd/documentam/security/SecurityUtilsTest.java (hunk 11 above)
Comment (smathangi) [reply src/test/java/uk/gov/hmcts/reform/ccd/documentam/security/SecurityUtilsTest.java:40]: it has `MockitoAnnotations.initMocks(this)` . I was a bit lazy to refactor this to use extension annotation.. (that annotation is very strict about unnecessary stubbings)

Comment (smathangi) [reply src/test/java/uk/gov/hmcts/reform/ccd/documentam/security/SecurityUtilsTest.java:35]: done now :)

File: build.gradle (hunk 3 above)
Comment (smathangi) [reply build.gradle:257]: i guess we are maintaining that as a separate property if multiple libraries from that group sharing same version.

Comment (sullivangareth) [reply build.gradle:257]: 👍

File: src/test/java/uk/gov/hmcts/reform/ccd/documentam/security/SecurityUtilsTest.java (hunk 11 above)
Comment (sullivangareth) [reply src/test/java/uk/gov/hmcts/reform/ccd/documentam/security/SecurityUtilsTest.java:35]: 👍

**Review by sullivangareth — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/apihelper/Constants.java [hunk 12]
```diff
@@ -39,16 +36,19 @@ private Constants() {
...
-    public static final String EXCEPTION_ERROR_ON_DOCUMENT_MESSAGE = "Exception occurred with operation on document id: %s";
+    public static final String EXCEPTION_ERROR_ON_DOCUMENT_MESSAGE = "Exception occurred with operation on document "
+        + "id: %s";
     public static final String EXCEPTION_ERROR_MESSAGE = "Exception occurred with operation";
     public static final String CASE_ID = "caseId";
     public static final String CASE_TYPE_ID = "caseTypeId";
     public static final String JURISDICTION_ID = "jurisdictionId";
-    public static final String SERVICE_PERMISSION_ERROR = "Service doesn't have sufficient permission on requested API {}";
+    public static final String SERVICE_PERMISSION_ERROR = "Service doesn't have sufficient permission on requested "
+        + "API {}";
```
Comment (tomxelliott) [inline src/main/java/uk/gov/hmcts/reform/ccd/documentam/apihelper/Constants.java:51]: Are we mixing String formatting approaches having both `%s` and `{}`?

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java [hunk 13]
```diff
@@ -1,90 +1,98 @@
...
-        final ProviderManager authenticationManager = (ProviderManager) authenticationManager();
-        authenticationManager.setEraseCredentialsAfterAuthentication(false);
-        authCheckerFilter.setAuthenticationManager(authenticationManager());
-
+    protected void configure(HttpSecurity http) throws Exception {
         http
-            .addFilter(authCheckerFilter)
+            .addFilterBefore(serviceAuthFilter, BearerTokenAuthenticationFilter.class)
             .sessionManagement().sessionCreationPolicy(STATELESS).and()
             .csrf().disable()
```
Comment (tomxelliott) [inline src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java:71]: This line is highlighted as a HIGH security risk by Sonar (https://sonarcloud.io/project/security_hotspots?id=ccd-case-document-am-api&hotspots=AXKlIBXrHtDcdm-qOd7q). Is there a simple solution to resolve this or should we park it for now?

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java [hunk 14]
```diff
@@ -1,90 +1,98 @@
...
-                    .encoder(new JacksonEncoder())
-                    .contract(new SpringMvcContract())
-                    .target(ServiceAuthorisationApi.class, s2sUrl);
+            .authenticated()
+            .and()
+            .oauth2ResourceServer()
+            .jwt()
+            .jwtAuthenticationConverter(jwtAuthenticationConverter)
+            .and()
+            .and()
```
Comment (tomxelliott) [inline src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java:82]: Do we need the 2 chained `and()` calls?

**Review by tomxelliott — DISMISSED** [review]
Looks good to me. Just a couple of minor comments

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java (hunk 14 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java:82]: yes

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java (hunk 13 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/ccd/documentam/configuration/SecurityConfiguration.java:71]: Need to look into that later> It's same security config we have in our other apps. Probably should address across ccd.

File: Jenkinsfile_CNP [hunk 15]
```diff
@@ -50,19 +43,14 @@ static LinkedHashMap<String, Object> secret(String secretName, String envVar) {
...
-
 // Vars needed for functional and smoke tests run against AKS
 env.IDAM_URL = "https://idam-api.aat.platform.hmcts.net"
 env.S2S_URL = "http://rpe-service-auth-provider-aat.service.core-compute-aat.internal"
 env.OAUTH2_CLIENT_ID = "ccd_gateway"
 env.OAUTH2_REDIRECT_URI = "https://www-ccd.aat.platform.hmcts.net/oauth2redirect"
 env.DOCUMENT_STORE_URL = "http://dm-store-aat.service.core-compute-aat.internal"
 env.DEFINITION_STORE_HOST = "http://ccd-definition-store-api-pr-575.service.core-compute-preview.internal"
-env.CCD_DATA_STORE_URL = "http://ccd-data-store-api-pr-601.service.core-compute-preview.internal"
+env.CCD_DATA_STORE_URL = "http://ccd-data-store-api-pr-1260.service.core-compute-preview.internal"
```
[the remark this answers is not in the document]
Comment (smathangi) [reply Jenkinsfile_CNP:53]: yeah.. that's good idea.. Lets in sync with other repos (aac)

File: src/main/java/uk/gov/hmcts/reform/ccd/documentam/apihelper/Constants.java (hunk 12 above)
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/ccd/documentam/apihelper/Constants.java:51]: probably something we have to fix that based on usage. Just fixed check-style error (line length?)

**Review by sullivangareth — APPROVED** [review]

**Review by tomxelliott — APPROVED** [review]
Updates look good to me 👍

**Pull request merged**